src/codegen/compile.ml

```diff
@@ -3697,6 +3727,148 @@ module IC = struct
+          Stack.with_words env "temp_ptr" 1l (fun get_temp_ptr ->
+            get_temp_ptr ^^ get_value ^^ store_unskewed_ptr ^^
+            get_offset ^^
+            get_temp_ptr ^^ compile_unboxed_const 4l ^^
+            IC.system_call env "ic0" "stable_write"))
+    | _ -> assert false
+
+
+  (* read and clear word32 from stable mem offset on stack *)
+  let read_and_clear_word32 env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code1 env "__stablemem_read_word32"
+        ("offset", I32Type) [I32Type]
+        (fun env get_offset ->
+          Stack.with_words env "temp_ptr" 1l (fun get_temp_ptr ->
+            let (set_word, get_word) = new_local env "word" in
+            (* read word *)
+            get_temp_ptr ^^ get_offset ^^  compile_unboxed_const 4l ^^
+            IC.system_call env "ic0" "stable_read" ^^
+            get_temp_ptr ^^ load_unskewed_ptr ^^
+            set_word ^^
+            (* write 0 *)
+            get_temp_ptr ^^ compile_unboxed_const 0l ^^ store_unskewed_ptr ^^
+            get_offset ^^
+            get_temp_ptr ^^ compile_unboxed_const 4l ^^
+            IC.system_call env "ic0" "stable_write" ^^
+            (* return word *)
+            get_word
+        ))
+    | _ -> assert false
+
+  (* grow (real) stable memory if needed (WIP)*)
+  let grow env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code1 env "__stablemem_grow"
+        ("pages", I32Type) []
+        (fun env get_pages ->
+          let (set_pages_needed, get_pages_needed) = new_local env "pages_needed" in
+          get_pages ^^
+          E.call_import env "ic0" "stable_size" ^^
+          G.i (Binary (Wasm.Values.I32 I32Op.Sub)) ^^
+          set_pages_needed ^^
+
+          get_pages_needed ^^
+          compile_unboxed_zero ^^
+          G.i (Compare (Wasm.Values.I32 I32Op.GtS)) ^^
+          G.if_ []
+            ( get_pages_needed ^^
+              E.call_import env "ic0" "stable_grow" ^^
+              (* Check result *)
+              compile_unboxed_zero ^^
+              G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^
+              E.then_trap_with env "Cannot grow stable memory."
+            ) G.nop)
+    | _ -> assert false
+
+  (* ensure stable memory includes [offset..offset+size) *)
+  let ensure env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code2 env "__stablemem_ensure"
+        (("offset", I32Type), ("size", I32Type)) []
+        (fun env get_offset get_size ->
+          get_offset ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+          get_size ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+          G.i (Binary (Wasm.Values.I64 I64Op.Add)) ^^
+          compile_const_64 (Int64.of_int page_size_bits) ^^
+          G.i (Binary (Wasm.Values.I64 I64Op.ShrU)) ^^
+          G.i (Convert (Wasm.Values.I32 I64Op.WrapI64)) ^^ (* TBR *)
+          compile_add_const 1l ^^
```

Comment:
Why add 1?

Reply:
grow takes the desired page *count* and really grows if necessary. But maybe I'm wrong. Will double check.

Reply:
I think there is an off by one earlier, since we want the address offset+size-1 to be accessible, not offset+size.

Reply:
The add 1 on this line adds an entire page, right? So this seems to imply that `ensure` always grows by at least one page. Should it really?

Reply:
ensure only grows if necessary, so it shouldn't, no. 
Also,the code assumes size > 0, so at least one page will be necessary.